### PR TITLE
Universal/shadow smooth fade

### PR DESCRIPTION
**Pull request by Nzollner**
### Purpose of this PR
This PR is implementing the smooth shadow fading.
For now, the smooth shadow fading is always turned on until we get a proper way to turn on/off the smooth shadow fading feature.

---
### Testing status
I tested the feature locally on my machine in the template scene for URP.
I ran the tests locally and checked if the result was correct.

**Automated Tests**:
I added a new test scene 131_ShadowFading to test the new feature.

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
I haven't written any documentation for this feature but I will write it when the 2. part of this PR is pushed also handling were to change the setting whether or not to use smooth shadow fading.

### CODE REVIEW

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl [hunk 1]
```diff
@@ -288,6 +288,17 @@ float3 ApplyShadowBias(float3 positionWS, float3 normalWS, float3 lightDirection
...
 
+float ApplyShadowFade(float shadowAttenuation, float3 positionWS)
+{
+    float3 camToPixel = positionWS - _WorldSpaceCameraPos;
+    float distanceCamToPixel2 = dot(camToPixel, camToPixel);
+    float shadowDist = _MainLightShadowParams.z;
+    float startShadowFadeDist = shadowDist * 0.9;
+
+    float fade = saturate((distanceCamToPixel2 - startShadowFadeDist) / (shadowDist - startShadowFadeDist));
+    return shadowAttenuation + (1 - shadowAttenuation) * fade;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:299]: quadratic falloff here should give more natural results.

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl [hunk 2]
```diff
@@ -288,6 +288,17 @@ float3 ApplyShadowBias(float3 positionWS, float3 normalWS, float3 lightDirection
...
 }
 
+float ApplyShadowFade(float shadowAttenuation, float3 positionWS)
+{
+    float3 camToPixel = positionWS - _WorldSpaceCameraPos;
+    float distanceCamToPixel2 = dot(camToPixel, camToPixel);
+    float shadowDist = _MainLightShadowParams.z;
+    float startShadowFadeDist = shadowDist * 0.9;
+
+    float fade = saturate((distanceCamToPixel2 - startShadowFadeDist) / (shadowDist - startShadowFadeDist));
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl:298]: only variable here is distanceCamToPixel2, this expression could be optimized to fit a single MAD instruction (Mul + Add) like following:

(distanceCamToPixel - startShadowFadeDist) / (shadowDist - startShadowFadeDist)

distanceCamToPixel / (shadowDist - startShadowFadeDist) - startShadowFadeDist / (shadowDist - startShadowFadeDist)

this becomes:
distanceCamToPixel / A - B, where A and B are constants
we can store 1/A and -B in uniform and then the expression becomes:

distanceCamToPixel2 * OneOverA + MinusB (single MAD)

We could store these in _MainLightShadowParams.z and _MainLightShadowParams.w, but it's a little bit weird these are used for additional lights as well. Do we have already another generic constant to pack those?

Comment (Nzollner) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl]: I can't think of any other generic constants to pack them into which would fit better

File: com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl (hunk 1 above)
Comment (Nzollner) [reply com.unity.render-pipelines.universal/ShaderLibrary/Shadows.hlsl]: I have tried to implement it in the new version for a faster falloff

**Comment by phi-lira** [comment]
@Nzollner ABV is not green.

**Comment by Nzollner** [comment]
Now the Universal test succeeds but I have changed the reference images such that test 35 passes with the new changes.

**Comment by Nzollner** [comment]
I have checked the shadow fading when no cascades are are used and it seemed to work correctly.
Is there a specific scene where there is an error?

**Review by ellioman — APPROVED** [review]
Looks fine code wise. 
Let's be sure that changing the Shadow Distance in the URP Asset doesn't break other tests and get QA to verify everything is good. Let's see if QA can reproduce what Andre wrote or not. 
Looks good code wise.

**Comment by Nzollner** [comment]
Using no cascades should work properly now.

**Review by simon-engelbrecht-soerensen — APPROVED** [review]
Could not reproduce issue with it not working without cascades, so it seems to be fixed.
Tested with all API'S on PC/Mac and builds. Works as expected, with the caveat of it not working when the view distance is super short and angled in a specific way, but this is accepted for this PR.

**Pull request merged**